### PR TITLE
rust: fix returned Result type in getrandom

### PR DESCRIPTION
**Pull request by bonifaido**
Otherwise two different Result types are proposed for returning.

### CODE REVIEW

**Comment by bjorn3** [comment]
What is exactly the issue? Why should a too large buffer result in a panic rather than an error?

**Comment by ojeda** [comment]
Also, please read `Documentation/process/submitting-patches.rst`, e.g. the commit needs a signature, and the commit message should explain the rationale behind the changes.

**Comment by bonifaido** [comment]
> What is exactly the issue? Why should a too large buffer result in a panic rather than an error?

The issue is with `dest.len().try_into()?`, so no buffer overflow, just a simple cast and try_into returns a different Result than the host method. The same way as it is used in line 47.

I will fixup the PR according to the docs, sorry for the laziness.

**Comment by bjorn3** [comment]
> The issue is with dest.len().try_into()?, so no buffer overflow, just a simple cast

It casts from usize to i32 (int on the C side) which can overflow. In case of such an overflow `try_into()` returns an error. I think this error should be forwarded to the caller rather than cause an abort.

> and try_into returns a different Result than the host method.

Why is this an issue?

> The same way as it is used in line 47.

Adding entropy is an infallible operation unlike getting random data. If the entropy buffer is too large that is likely a bug and thus it makes sense to panic (implemented as kernel oops). If you need to get too much random data that may very well be caused by a userspace program. In that case you need to return an error and not panic.

**Comment by bonifaido** [comment]
Sure, I totally agree (didn't know all the details about a possible overflow here during the cast). My original issue without my changes is this:

```
  RUSTC L rust/kernel.o
error[E0277]: `?` couldn't convert the error to `error::Error`
  --> rust/kernel/random.rs:24:34
   |
15 | pub fn getrandom(dest: &mut [u8]) -> Result {
   |                                      ------ expected `error::Error` because of this
...
24 |             dest.len().try_into()?,
   |                                  ^ the trait `From<Infallible>` is not implemented for `error::Error`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = help: the following implementations were found:
             <error::Error as From<CStrConvertError>>
             <error::Error as From<LayoutError>>
             <error::Error as From<TryFromIntError>>
             <error::Error as From<TryReserveError>>
           and 3 others
   = note: required because of the requirements on the impl of `FromResidual<core::result::Result<Infallible, Infallible>>` for `core::result::Result<(), error::Error>`

error: aborting due to previous error
```

**Comment by alex** [comment]
What platform are you compiling for? This looks to be something where the cast is unexpectedly to the same integer size?

That said, the right fix would be to `impl From<Infallible> for Error`

**Comment by bonifaido** [comment]
I'm compiling on a 64bit x86 Ubuntu on VirtualBox:

`Linux osboxes 5.13.0-35-generic #40-Ubuntu SMP Mon Mar 7 08:03:10 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`

File: rust/kernel/error.rs [hunk 1]
```diff
@@ -420,6 +420,12 @@ impl From<core::fmt::Error> for Error {
     }
 }
 
+impl From<core::convert::Infallible> for Error {
+    fn from(_: core::convert::Infallible) -> Error {
+        unreachable!()
```
Comment (bjorn3) [inline rust/kernel/error.rs]: You can match on the `Infallible` value. It is an enum with no variants, so you don't need match arms. All (that is all 0) arms return `Error` (or any other type for that matter) so the match itself returns this too. This saves a tiny bit of code to be processed by the compiler.

Comment (alex) [inline rust/kernel/error.rs]:
```suggestion
    fn from(e: core::convert::Infallible) -> Error {
        match e {}
```

**Review by bjorn3 — APPROVED** [review]
Please squash.

**Review by alex — APPROVED** [review]

File: rust/kernel/error.rs (hunk 1 above)
Comment (bonifaido) [reply rust/kernel/error.rs]: Some new techniques I learned today, thanks @bjorn3 @alex !

**Pull request merged**